packages/task/src/node/task-server.ts

```diff
@@ -0,0 +1,233 @@
+        @inject(TaskManager) protected readonly taskManager: TaskManager,
+        @inject(TaskFactory) protected readonly taskFactory: TaskFactory
+    ) {
+        taskManager.onDelete(id => {
+            const toDispose = this.tasksToDispose.get(id);
+            if (toDispose !== undefined) {
+                toDispose.dispose();
+                this.tasksToDispose.delete(id);
+            }
+        });
+    }
+
+    dispose() {
+        // do nothing
+    }
+
+    getTasks(context?: string | undefined): Promise<TaskInfo[]> {
+        const taskinfo: TaskInfo[] = [];
+
+        const tasks = this.taskManager.getTasks(context);
+        if (tasks !== undefined) {
+            for (const task of tasks) {
+                taskinfo.push(task.getRuntimeInfo());
+            }
+        }
+        this.logger.debug(`getTasks(): about to return task information for ${taskinfo.length} tasks`);
+
+        return Promise.resolve(taskinfo);
+    }
+
+    async run(options: TaskOptions, ctx?: string): Promise<TaskInfo> {
+        // on windows, prefer windows-specific options, if available
+        const processOptions = (isWindows && options.windowsProcessOptions !== undefined) ?
+            options.windowsProcessOptions : options.processOptions;
+
+        const command = processOptions.command;
+
+        // sanity check - we expect the cwd to be set by the client.
+        if (!options.cwd) {
+            return Promise.reject(new Error("Can't run a task when 'cwd' is not provided by the client"));
+        }
+
+        const cwd = FileUri.fsPath(options.cwd);
+        // Use task's cwd with spawned process and pass node env object to
+        // new process, so e.g. we can re-use the system path
+        processOptions.options = {
+            cwd: cwd,
+            env: process.env
+        };
+
+        // When we create a process to execute a command, it's difficult to know if it failed
+        // because the executable or script was not found, or if it was found, ran, and exited
+        // unsuccessfully. So here we look to see if it seems we can find a file of that name
+        // that is likely to be the one we want, before attemting to execute it.
+        const cmd = await this.findCommand(command, cwd);
+        if (cmd) {
+            try {
+                // use terminal or raw process
+                let task: Task;
+                let proc: TerminalProcess | RawProcess;
+
+                if (options.processType === 'terminal') {
+                    this.logger.info('Task: creating underlying terminal process');
+                    proc = this.terminalProcessFactory(<TerminalProcessOptions>processOptions);
+                } else {
+                    this.logger.info('Task: creating underlying raw process');
+                    proc = this.rawProcessFactory(<RawProcessOptions>processOptions);
+                }
+
+                task = this.taskFactory(<TaskProcessOptions>
```

Comment:
Don't cast it here, if you do you lose compiler warnings

Reply:
cast removed - thanks: it let the compiler complain about some potential issue with optional parameters!